gitCoreJGit/src/main/java/com/virtuslab/gitcore/impl/jgit/GitCoreRepository.java

```diff
@@ -149,24 +152,50 @@ private boolean isBranchMissing(String path) throws GitCoreException {
         .isEmpty();
   }
 
-  @Override
-  public boolean isAncestor(BaseGitCoreCommit presumedAncestor, BaseGitCoreCommit presumedDescendant)
-      throws GitCoreException {
+  @Nullable
+  private RevCommit deriveMergeBase(BaseGitCoreCommit a, BaseGitCoreCommit b) throws GitCoreException {
     RevWalk walk = new RevWalk(jgitRepo);
-    walk.sort(RevSort.TOPO);
+    walk.setRevFilter(RevFilter.MERGE_BASE);
     try {
-      ObjectId descendantObjectId = jgitRepo.resolve(presumedDescendant.getHash().getHashString());
+      ObjectId descendantObjectId = jgitRepo.resolve(a.getHash().getHashString());
       assert descendantObjectId != null : "Cannot find descendant";
 
-      ObjectId ancestorObjectId = jgitRepo.resolve(presumedAncestor.getHash().getHashString());
+      ObjectId ancestorObjectId = jgitRepo.resolve(b.getHash().getHashString());
       assert ancestorObjectId != null : "Cannot find ancestor";
 
       walk.markStart(walk.parseCommit(descendantObjectId));
-
-      return Iterator.ofAll(walk).find(revCommit -> revCommit.getId().equals(ancestorObjectId)).isDefined();
-
+      walk.markStart(walk.parseCommit(ancestorObjectId));
+      return walk.next();
     } catch (IOException e) {
       throw new GitCoreException(e);
     }
   }
+
+  private final java.util.Map<Tuple2<BaseGitCoreCommit, BaseGitCoreCommit>, @Nullable GitCoreCommitHash> mergeBaseCache = new java.util.HashMap<>();
+
+  @Nullable
+  private GitCoreCommitHash deriveMergeBaseIfNeeded(BaseGitCoreCommit a, BaseGitCoreCommit b) throws GitCoreException {
+    var abKey = Tuple.of(a, b);
+    var baKey = Tuple.of(b, a);
+    if (mergeBaseCache.containsKey(abKey)) {
+      return mergeBaseCache.get(abKey);
```

Comment:
I think it is worth to notice here that:
https://stackoverflow.com/questions/4553624/hashmap-get-put-complexity

Of course, it would a minor improvement (supposing that no one changes `Map` implementation) but instead of `containsKey` we could just compare the `get` result with `null` (and reduce the complexity twice)

Reply:
Hmm the problem is that specific map is that values are `@Nullable` (to handle rare cases when merge-base can't be found for a specific pair of commits)... so `null` is a rare-yet-valid value and isn't equivalent to a missing key :/